torch/_inductor/ir.py

```diff
@@ -2332,28 +2332,55 @@ def realize_into(cls, src, dst):
         # NOTE: We must realize users of `dst` before we realize `src`, since
         # realization order determines scheduling order. Otherwise, src's
         # mutation would be scheduled before the existing users of dst!
-        V.graph.mark_buffer_mutated(dst.get_name())
+        realized_bufs = V.graph.mark_buffer_mutated(dst.get_name())
 
         if isinstance(src, TensorBox):
             src = src.data
 
-        # We copy the contents of src into dst. In most cases this should
-        # be fused into a single kernel by the scheduler.
-        # NOTE: We cannot change src's layout to mutate dst directly as this
-        # would alias src to dst, which is not correct as further mutations to
-        # dst would effect users of src.
-        src.realize_hint()
+        need_copy = True
 
-        src = Pointwise.create(
-            device=src.get_device(),
-            dtype=src.get_dtype(),
-            inner_fn=src.make_loader(),
-            ranges=[
-                V.graph.sizevars.guard_equals(a, b)
-                for a, b in zip(src.get_size(), dst.get_size())
-            ],
-        ).data
-        src.realize()
+        if (
+            isinstance(src.data, Buffer)
+            and isinstance(src.get_layout(), FlexibleLayout)
+            and not src.is_zero_elements()
+            and (
+                (src.get_name() in realized_bufs)
```

Comment:
If `src` was realized doesn't that imply `src` is a user of `dst` and so copying is necessary in case the indexing doesn't line up. For example something like `x.copy_(x.transpose())` is unlikely to work.

Reply:
Thanks for the comments.

- `x.copy_(x.transpose())` may not be a good example. Seems it will also fail in eager mode. And 2 tensors with different sizes may not be copied.

- Another way comes to my mind may be as the complement of https://github.com/pytorch/pytorch/pull/107614
If src will be realized here, https://github.com/pytorch/pytorch/blob/e942fddb83c0a101e0bb10fd66538cf58254bd5f/torch/_inductor/ir.py#L2368
Can we skip the src realization in https://github.com/pytorch/pytorch/blob/e942fddb83c0a101e0bb10fd66538cf58254bd5f/torch/_inductor/ir.py#L2347
I tried this idea, it can also fix the regression in https://github.com/pytorch/pytorch/issues/109874

Reply:
> `x.copy_(x.transpose())` may not be a good example. Seems it will also fail in eager mode.

Okay, the same principle applies to `x.transpose().clone()`.

>  Can we skip the src realization in [mark_buffer_mutated]

No I don't think this is sound. If `src` is a user of `dst` and `src` isn't realized before the mutation then other users of `src` may end up doing the calculation with the post-mutation version of `dst`.